src/ore/src/tracing/optimizer.rs

```diff
@@ -0,0 +1,313 @@
+///     construct ad-hoc spans (see the iteration spans in the `Fixpoint`
+///     transform for example).
+///
+/// As a consequence of the second constraint, a sequence of paths such as
+/// ```text
+/// optimizer.foo.bar
+/// optimizer.foo.baz
+/// ```
+/// is not well-formed as it is missing the results of the prefix paths at the
+/// end:
+/// ```text
+/// optimizer.foo.bar
+/// optimizer.foo.baz
+/// optimizer.foo
+/// optimizer
+/// ```
+///
+/// Also, note that full paths can be repeated within a pipeline, but adjacent
+/// duplicates are interpreted as separete invocations. For example, the
+/// sub-sequence
+/// ```text
+/// ... // preceding stages
+/// optimizer.foo.bar // 1st call
+/// optimizer.foo.bar // 2nd call
+/// ... // following stages
+/// ```
+/// will be rendered by the UI as the following tree structure.
+/// ```text
+/// optimizer
+///   ... // following stages
+///   foo
+///     bar // 2nd call
+///     bar // 1st call
+///   ... // preceding stages
+/// ```
+pub fn trace_plan<T: Clone + 'static>(plan: &T) {
+    tracing::Span::current().with_subscriber(|(_id, subscriber)| {
+        if let Some(trace) = subscriber.downcast_ref::<PlanTrace<T>>() {
+            // Take the current time before cloning the plan to exclude the clone()
+            // time from the duration computed by `push` when creating a new TraceEntry.
+            let current_time = std::time::Instant::now();
+            let plan = plan.clone();
+            trace.push(plan, current_time)
+        }
+    });
+}
+
+/// A [`layer::Layer`] implementation for [`PlanTrace`].
+///
+/// Populates the `data` wrapped by the [`PlanTrace`] instance with
+/// [`TraceEntry`] values, one for each span with attached plan in its
+/// extensions map.
+impl<S, T> layer::Layer<S> for PlanTrace<T>
+where
+    S: subscriber::Subscriber + for<'span> tracing_subscriber::registry::LookupSpan<'span>,
+    T: 'static,
+{
+    fn on_new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, ctx: layer::Context<'_, S>) {
+        // push to path stack
+        let mut path = self.path.lock().expect("path shouldn't be poisoned");
+        let name = attrs
+            .get_str("path.segment")
+            .unwrap_or_else(|| ctx.span(id).expect("span").name().to_string());
+        path.deref_mut().push(name);
+    }
+
+    fn on_enter(&self, _id: &span::Id, _ctx: layer::Context<'_, S>) {
+        // push to time stack
+        let mut times = self.times.lock().expect("times shouldn't be poisoned");
+        times.deref_mut().push(std::time::Instant::now());
+    }
+
+    fn on_exit(&self, _id: &span::Id, _ctx: layer::Context<'_, S>) {
+        // pop from path stack
+        let mut path = self.path.lock().expect("path shouldn't be poisoned");
+        path.deref_mut().pop();
+        // pop from time stack
+        let mut times = self.times.lock().expect("times shouldn't be poisoned");
+        times.deref_mut().pop();
```

Comment:
is there only one time stack for the whole optimization (well, per-`T`)

Reply:
Also, a comment explaining how this works would be nice, when the span is exited (not closed), we pop from the path stack; what actually reads from the path stack?

Reply:
Oh I see, its `current_path`

I believe this will all be cleaned up if we use `extensions_mut` instead

Reply:
> is there only one time stack for the whole optimization

Yes, there is one time and one path stack conceptually, and for lack of further optimizing this at the moment we are going to maintain one copy per plan type `T`.